Lab 5 ibl

MChepulis opened the pull request.

MChepulis committed:
add lab_5 project

just a copy of lab_4

MChepulis committed:
replace compute of irradiance map to initialize

Zhavoronkova-Alina committed:
Create prefilteered color texture

Add prefiltered color PS
Create prefilteered_color texture
Create prefilteered_color resource_view

Zhavoronkova-Alina committed:
Add ibl to brdf shader

Add preintegarted brdf PS
Create prefilteered_color texture
Create prefilteered_color resource_view
Update brdf PS
Add enum of default albedo colors

MChepulis committed:
refactor Graphics

refactor Graphics.h
add default ambient color

refactor Graphics.cpp
add prefiltered color
add preintegrated brdf

MChepulis committed:
set c++17

MChepulis committed:
fix remarks

1) limited dot(n, v) from below in G_SCHLICKGGX
2) fix the center of the env. sphere at the camera position
3) now roughness in [0.001, 1]
4) add copy texture files (.dds) to target dir during build

anthark reviewed (changes requested):
Есть проблема с мерджем этой ветки с главной, надо поправить. И, насколько я могу судить, есть ошибка компиляции на x64, а у нас именно эта платформа основная. 

И самое главное - выглядит так, что roughness не работает с IBL: все столбцы шариков выглядят одинаково, а на самом деле отражение должно постепенно становится все более мутным по мере продвижения от левого к правому, пока не исчезнет совсем.

MChepulis committed:
fix remark for Lab_5 (not all)

fix compile error for x64 built
add mips to env_sphere_texture

FIXME: read prefiltered color only from mip 0

MChepulis committed:
fix zero-mip problem

add mip levels to ShaderResourceView

AntiFrizz1 committed:
Merge remote-tracking branch 'origin/main' into Lab_5-ibl

anthark reviewed (approved):
Хорошо, работает. Задача засчитана.
